src/vision/status_quo/barbara_gets_burned_by_select.md

```diff
@@ -0,0 +1,76 @@
+
+If you would like to expand on this story, or adjust the answers to the FAQ, feel free to open a PR making edits (but keep in mind that, as they reflect peoples' experiences, status quo stories [cannot be wrong], only inaccurate). Alternatively, you may wish to [add your own status quo story][htvsq]!
+
+## The story
+
+Barbara is working on the [YouBuy] server. In one particular part of the story, she has a process that has to load records from a database on the disk. As she receives data from the database, the data is sent into a channel for later processing. She writes an `async fn`  that looks something like this:
+
+```rust
+async fn read_send(db: &mut Database, channel: &mut Sender<...>) {
+  loop {
+    let data = read_next(db).await;
+    let items = parse(&data);
+    for item in items {
+      channel.send(item).await;
+    }
+  }
+}
+```
+
+This database load has to take place while also fielding requests from the user. The routine that invokes `read_send` uses [`select!`](https://docs.rs/futures/0.3.14/futures/macro.select.html) for this purpose. It looks something like this:
+
+```rust
+let mut db = ...;
+let mut channel = ...;
+loop {
+    futures::select! {
+        _ = read_send(&mut file, &mut channel) => {},
+        some_data = socket.read_packet() => {
+            // ...
+        }
+    }
+}
+```
+
+This setup seems to work well a lot of the time, but Barbara notices that the data getting processed is sometimes incomplete. It seems to be randomly missing some of the rows from the middle of the database, or individual items from a row.
+
+### Debugging
+
+She's not sure what could be going wrong! She starts debugging with print-outs and logging. Eventually she realizes the problem. Whenever a packet arrives on the socket, the `select!` macro will drop the other futures. This can sometime cause the `read_send` function to be canceled in between reading the data from the disk and sending the items over the channel. Ugh!
+
+Barbara has a hard time figuring out the best way to fix this problem.
+
+## 🤔 Frequently Asked Questions
+
+### **What are the morals of the story?**
+
+* Cancellation doesn't always cancel the entire task; particularly with `select!`, it sometimes cancels just a small piece of a given task.
+    * This is in tension with Rust's original design, which was meant to tear down an entire thread or task at once, precisely because of the challenge of writing exception-safe code.
+* Writing "cancellation safe" code is very challenging.
+
+### **What are the sources for this story?**
+
+This was based on [tomaka's blog post](https://tomaka.medium.com/a-look-back-at-asynchronous-rust-d54d63934a1c), which also includes a number of possible solutions, all of them quite grungy.
+
+### **Why did you choose Barbara to tell this story?**
+
+tomaka is a veteran Rust user.
```

Comment:
Should we put "tomoka" here which is not within the 4 characters?

Reply:
It's fine by me to either mention or not mention me, whatever you decide